### ISSUE TITLE
Authenticated session expires when initial jwt token expires

### ISSUE THREAD

**Issue by absoludity** [kubeapps/kubeapps#2686]
### Description:

Noticed while testing on TKG that when we login via OIDC (of any type), the session lasts only until the token expires, whereas oauth2-proxy should be refreshing the token.

### Steps to reproduce the issue:

1. Login via OIDC
2. Leave the browser tab open for 5-10 mins (depending on the token time set in your oauth2 client settings)
3. Try to use Kubeapps

### Describe the results you received:

A helpful notice about the session expiring and logging out.

### Describe the results you expected:

oauth2-proxy should be renewing the token a certain number of times.

### Additional information you deem important (e.g. issue happens only occasionally):

It may be we just need to test using the [--cookie-refresh option](https://oauth2-proxy.github.io/oauth2-proxy/docs/configuration/overview/#command-line-options) to oauth2-proxy, not sure.

**Comment by dlaloue-vmware** [comment]
it looks like the refresh issue is a know issue in oauth2 proxy and is being worked on:
    https://github.com/oauth2-proxy/oauth2-proxy/issues/490
I will monitor that thread to know when a fix is available and try to verify it locally.

**Comment by absoludity** [comment]
I think that's a different issue? That describes the situation where once the id_token expires, oauth2-proxy does refresh to get a new token, but doesn't update the session age, and so continues to refresh the token unnecessarily from that point on (until the access token is also expired).

The situation we want to check here is whether the refresh even happens without logging the user out: my experience (which I just dumped into this issue) was that I was logged out as soon as the first id_token expired, rather than being able to continue with a refreshed token. It could just be the configuration, not sure, but we need to ensure that when the id token expires, oauth2-proxy refreshes and provides a new one transparently to Kubeapps, without kubeapps logging out etc.

**Comment by antgamdia** [comment]
I've been having a look at the issue as well:

## Configurations:

### Case 1: access token expires with a proper refresh. Expected result: auto-refresh, no logout :ok: 

```
Access token TTL: 1 min
Refresh token TTL: 60 days
--cookie-refresh=10s
```

``` 
[2021/05/24 10:18:44] [oidc.go:158] refreshed session: Session{email:user@example.com user:example.com:f43ee241-e1b1-4bde-a4b2-b894a08bf601 PreferredUsername: token:true id_token:true created:2021-05-24 10:18:44.183820965 +0000 UTC m=+3001.435142281 expires:2021-05-24 10:19:43.183304267 +0000 UTC m=+3060.434625483 refresh_token:true}
[2021/05/24 10:18:44] [oidc.go:158] refreshed session: Session{email:user@example.com user:example.com:f43ee241-e1b1-4bde-a4b2-b894a08bf601 PreferredUsername: token:true id_token:true created:2021-05-24 10:18:44.205138763 +0000 UTC m=+3001.456459979 expires:2021-05-24 10:19:43.204778165 +0000 UTC m=+3060.456099481 refresh_token:true}
[2021/05/24 10:18:44] [session_store.go:164] WARNING: Multiple cookies are required for this session as it exceeds the 4kb cookie limit. Please use server side session storage (eg. Redis) instead.
[2021/05/24 10:18:44] [oidc.go:158] refreshed session: Session{email:user@example.com user:example.com:f43ee241-e1b1-4bde-a4b2-b894a08bf601 PreferredUsername: token:true id_token:true created:2021-05-24 10:18:44.339490623 +0000 UTC m=+3001.590811939 expires:2021-05-24 10:19:43.339027526 +0000 UTC m=+3060.590348842 refresh_token:true}
[2021/05/24 10:18:44] [session_store.go:164] WARNING: Multiple cookies are required for this session as it exceeds the 4kb cookie limit. Please use server side session storage (eg. Redis) instead.
10.244.2.242:43506 - 8fa5b5277a7b748618b5293c3976b85f - user@example.com [2021/05/24 10:18:43] mycluster.example.com  GET / "/api/v1/clusters/tmc-aws-2/namespaces/kubeapps/apprepositories" HTTP/1.1 "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/90.0.4430.212 Safari/537.36" 404 302 3.115
```


### Case 2: access token expires but slow refresh. Expected result: no auto-refresh, logout instead :ok: 

```
Access token TTL: 1 min
Refresh token TTL: 60 days
--cookie-refresh=1h
```

``` 
[stored_session.go:76] Error loading cookied session: cookie "_oauth2_proxy" not present, removing session
mycluster.example.com GET - "/api/clusters/tmc-aws-2/" HTTP/1.1 "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/90.0.4430.212 Safari/537.36" 401 0 0.000
``` 

### Case 3: refresh token expires. Expected result: no auto-refresh, logout instead :ok: 

```
Access token TTL: 1 min
Refresh token TTL: 1 min
--cookie-refresh=10s
```

```
[stored_session.go:122] Refreshing 8m2.528604298s old session cookie for Session{email:user@example.com user:example.com:f43ee241-e1b1-4bde-a4b2-b894a08bf601 PreferredUsername: token:true id_token:true created:2021-05-24 09:52:04.471395702 +0000 UTC expires:2021-05-24 09:53:03.470887904 +0000 UTC refresh_token:true} (refresh after 10s)
mycluster.example.com GET / "/favicon-32x32.png" HTTP/1.1 "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/90.0.4430.212 Safari/537.36" 304 0 0.003
stored_session.go:76] Error loading cookied session: error refreshing access token for session (Session{email:user@example.com user:example.com:f43ee241-e1b1-4bde-a4b2-b894a08bf601 PreferredUsername: token:true id_token:true created:2021-05-24 09:52:04.471395702 +0000 UTC expires:2021-05-24 09:53:03.470887904 +0000 UTC refresh_token:true}): error refreshing access token: unable to redeem refresh token: failed to get token: oauth2: cannot fetch token: 400 Bad Request
Response: {"requestId":"59c05bf2-6f8b-434d-ba2b-e5517cbcc92d","message":"invalid_grant: Invalid refresh token: xxxx...Rc2i8","metadata":null,"cspErrorCode":"340.206-340.800","moduleCode":340,"statusCode":400}, removing session
[stored_session.go:76] Error loading cookied session: error refreshing access token for session (Session{email:user@example.com user:example.com:f43ee241-e1b1-4bde-a4b2-b894a08bf601 PreferredUsername: token:true id_token:true created:2021-05-24 09:52:04.471395702 +0000 UTC expires:2021-05-24 09:53:03.470887904 +0000 UTC refresh_token:true}): error refreshing access token: unable to redeem refresh token: failed to get token: oauth2: cannot fetch token: 400 Bad Request
Response: {"metadata":null,"moduleCode":340,"statusCode":400,"message":"invalid_grant: Invalid refresh token: xxxx...Rc2i8","requestId":"d75015da-1a67-4fe5-97eb-a729ea5c1384","cspErrorCode":"340.206-340.800"}, removing session
mycluster.example.com GET - "/api/clusters/tmc-aws-2/" HTTP/1.1 "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/90.0.4430.212 Safari/537.36" 401 0 0.968
```

## Conclusions:

Apparently, there is no issue but a lack of configuration. Let's see:

Since oauth2proxy is not refreshing automatically the cookie, we have to manually add this config the additionalFlags; namely:

```yaml
authProxy:
  additionalFlags:
  - --cookie-refresh=30s
 ...
```

Ideas: 

- Add a default value in the chart, like 30 minutes (Google's default is 60 min, Dex 24h, Okta 30 days, Keycloak 60 days, Azure 1h)
- (and/or) Add this info in our docs (both general and the new step-by-step)




BTW @absoludity, I wanted to get rid of `insecure-oidc-skip-issuer-verification=true`, so I tried adding manually all the endpoints, as follows:

```yaml
  - --skip-oidc-discovery=true
  - --login-url=https://console-stg.cloud.vmware.com/csp/gateway/discovery
  - --oidc-issuer-url=https://gaz-preview.csp-vidm-prod.com
  - --login-url=https://console-stg.cloud.vmware.com/csp/gateway/discovery?orgLink=%2Fcsp%2Fgateway%2Fam%2Fapi%2Forgs%2F<REDACTED_ORGANIZATION_ID>
  - --redeem-url=https://console-stg.cloud.vmware.com/csp/gateway/am/api/auth/token
  - --oidc-jwks-url=https://console-stg.cloud.vmware.com/csp/gateway/am/api/auth/token-public-key?format=jwks
  - --profile-url=https://console-stg.cloud.vmware.com/csp/gateway/am/api/userinfo
```

I'm not seeing any error regarding the mismatch (well, there is no mismatch since we are not really relying on the oidc discovery). So, perhaps we can say this info in the docs.

**Comment by absoludity** [comment]
> ## Conclusions:
> 
> Apparently, there is no issue but a lack of configuration.

Excellent, best possible outcome.

> Let's see:
> 
> Since oauth2proxy is not refreshing automatically the cookie, we have to manually add this config the additionalFlags; namely:
> 
> ```yaml
> authProxy:
>   additionalFlags:
>   - --cookie-refresh=30s
>  ...
> ```
> 
> Ideas:
> 
>     * Add a default value in the chart, like 30 minutes (Google's default is 60 min, Dex 24h, Okta 30 days, Keycloak 60 days, Azure 1h)

A default value for what? For cookie refresh? It needs to be less than the access token TTL no? Or perhaps you mean the TTL for the IdP token (which would be set in the identity provider, not the chart?). I'm not sure exactly what you mean here.

> 
>     * (and/or) Add this info in our docs (both general and the new step-by-step)

+1

> 
> BTW @absoludity, I wanted to get rid of `insecure-oidc-skip-issuer-verification=true`, so I tried adding manually all the endpoints, as follows:
> 
> ```yaml
>   - --skip-oidc-discovery=true
>   - --login-url=https://console-stg.cloud.vmware.com/csp/gateway/discovery
>   - --oidc-issuer-url=https://gaz-preview.csp-vidm-prod.com
>   - --login-url=https://console-stg.cloud.vmware.com/csp/gateway/discovery?orgLink=%2Fcsp%2Fgateway%2Fam%2Fapi%2Forgs%2F<REDACTED_ORGANIZATION_ID>
>   - --redeem-url=https://console-stg.cloud.vmware.com/csp/gateway/am/api/auth/token
>   - --oidc-jwks-url=https://console-stg.cloud.vmware.com/csp/gateway/am/api/auth/token-public-key?format=jwks
>   - --profile-url=https://console-stg.cloud.vmware.com/csp/gateway/am/api/userinfo
> ```
> 
> I'm not seeing any error regarding the mismatch (well, there is no mismatch since we are not really relying on the oidc discovery). So, perhaps we can say this info in the docs.

Excellent - yeah, I'd only ever used the self-discovery which leads to the tls mis-match. If we can avoid that by recommending (for CSP) to skip the oidc discovery and just set them manually, that's much better.

**Comment by dlaloue-vmware** [comment]
I have been rerunning tests since this morning with different timeouts for the access token and cookie-refresh, but i have not been able to reproduce the behavior observed yesterday (and i deleted the log file i had saved, so i can't analyze it).

From today's logs, I can see the cookie refresh issue as mentioned, where the refreshed cookie is not returned in the response headers,  meaning oauth2 proxy will keep doing a cookie refresh over and over.

But this does not impact the UI as once the token is expired, there will be again a cookie refresh and this time the new cookie is returned in the response headers.

As was proposed, we should add the --cookie-refresh option with a default value.
I will try to lookup what the default values are for various providers to get an idea of the default value we could use. But it is safer to use a short value as default, and let users increase it based on their setup.

**Comment by dlaloue-vmware** [comment]
here are info i could find via google  (note that some docs refer to session in place of refresh) for default values:
  keycloak:   token: 5mn,  session 30mn idle / 10h max
  google:     token:  1hr
  dex:        token: 12h
  okta:       token: 60mn    session: 100days
  microsoft:   token: 10mn   session:  90days or unlimited
  IBM cloud provider:   token : 1h   refresh:  1-90 days
  Pivotal UAA:   token: 12h

so values for the access token / id token have a very wide range, from 5/10mn to 12h+

maybe something like 1mn or 2mn would be good at least for dev installs.
we should document that it needs to be tuned according to the target issuer.

**Comment by dlaloue-vmware** [comment]
This issue was fixed via a Chart changed made in version 7.1.0
